internal/util/kubernetes/kubernetes.go

```diff
@@ -133,3 +135,65 @@ func DeleteManagedObjects(ctx context.Context, client k8sClient.Client, objs map
 	}
 	return nil
 }
+
+// ownerExists checks if an owner reference already exists in the list of owner references.
+func ownerExists(existingRefs []interface{}, ownerRef map[string]interface{}) bool {
+	var alreadyExists bool
+	for _, ref := range existingRefs {
+		if refMap, ok := ref.(map[string]interface{}); ok {
+			if refMap["uid"] == ownerRef["uid"] {
+				alreadyExists = true
+				break
+			}
+		}
+	}
+	return alreadyExists
+}
+
+func ApplyOwnership(manifest string, controllerRollout *v1alpha1.NumaflowControllerRollout) ([]byte, error) {
+	// Decode YAML into an Unstructured object
+	decUnstructured := yamlserializer.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+	obj := &unstructured.Unstructured{}
+	_, _, err := decUnstructured.Decode([]byte(manifest), nil, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct the new owner reference
+	ownerRef := map[string]interface{}{
+		"apiVersion":         controllerRollout.APIVersion,
+		"kind":               controllerRollout.Kind,
+		"name":               controllerRollout.Name,
+		"uid":                string(controllerRollout.UID),
+		"controller":         true,
+		"blockOwnerDeletion": true,
+	}
+
+	// Get existing owner references and check if our reference is already there
+	existingRefs, found, err := unstructured.NestedSlice(obj.Object, "metadata", "ownerReferences")
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		existingRefs = []interface{}{}
+	}
+
+	// Check if the owner reference already exists to avoid duplication
+	alreadyExists := ownerExists(existingRefs, ownerRef)
+
+	// Add the new owner reference if it does not exist
```

Comment:
something to think about later: are we retrying updates to objects if there's a conflict in resourceVersion? (even if we're being thread safe internally, there's potential for somebody outside of numaplane to modify them)